### PR TITLE
Add simple benchmarks for JSON coding

### PR DESCRIPTION
**Pull request by Catfish-Man**
This is primitive, but it'll give us a baseline to work from

### CODE REVIEW

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test

File: benchmark/single-source/JSON.swift [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import TestsUtils
+import Foundation
+
+public let JSON = [
+  BenchmarkInfo(name: "JSONPerfEncode", runFunction: run_JSONPerfEncode, tags: [.validation, .bridging], setUpFunction: setup_coder),
+  BenchmarkInfo(name: "JSONPerfDecode", runFunction: run_JSONPerfDecode, tags: [.validation, .bridging], setUpFunction: setup_coder),
+]
+
+
+struct Tove : Codable {
```
Comment (lancep) [inline benchmark/single-source/JSON.swift]: Style nit: the `:` should not have a space to the left
```suggestion
struct Tove: Codable {
```

Comment (lancep) [reply benchmark/single-source/JSON.swift]: Same goes for the rest of the file

File: benchmark/single-source/JSON.swift [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  var encoder: JSONEncoder = JSONEncoder()
+  var decoder: JSONDecoder = JSONDecoder()
+  var data: Data! = nil
+  
+  init() {
+    data = try! encoder.encode(poems)
+  }
+}
+
+var tester:JSONPerfTester! = nil
```
Comment (lancep) [inline benchmark/single-source/JSON.swift]:
```suggestion
var tester: JSONPerfTester! = nil
```

File: benchmark/single-source/JSON.swift [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  var wabe = Wabe.canonical
+  var beware: [Beast] = [ Beast(name: "Jabberwock"), Beast(name: "Bandersnatch"), Beast(name: "Jubjub Bird")]
+  var swordType = "vörpal"
+}
+
+struct JSONPerfTester {
+  var poems = Array(repeating: Jabberwocky(), count: 6)
+  var encoder: JSONEncoder = JSONEncoder()
+  var decoder: JSONDecoder = JSONDecoder()
+  var data: Data! = nil
```
Comment (itaiferber) [inline benchmark/single-source/JSON.swift]: Is there a reason to give this a default value?

**Review by itaiferber — COMMENTED** [review]
Looks like a reasonable start to me. For evolving this in the future, though: what do we directly expect to measure here? Since `JSONEncoder`/`JSONDecoder` rely on `JSONSerialization`, we'll also be picking up any fluctuations there over time; is that acceptable for these benchmarks? Are the tests here intending to measure the cost of the JSON side of things, the `Codable` side of things, or both?

(I ask the latter because I've long wanted to introduce a base `Encoder`/`Decoder` that does nothing but collect an object graph into Swift collections without encoding to any format — ideally allowing us to unit-test/benchmark the `Codable` side of things instead of the underlying format.)

File: benchmark/single-source/JSON.swift (hunk 3 above)
Comment (Catfish-Man) [reply benchmark/single-source/JSON.swift]: Probably not

**Comment by Catfish-Man** [comment]
@itaiferber I wanted an end-to-end test so that we can catch regressions on the OS side as well (and just because that's what users actually care about). I agree it would be good to have tests for the individual pieces as well.

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test

File: benchmark/single-source/JSON.swift [hunk 4]
```diff
@@ -0,0 +1,92 @@
+//===--- JSON.swift ------------------------------------===//
```
Comment (moiseev) [inline benchmark/single-source/JSON.swift]:
```suggestion
//===--- JSON.swift -------------------------------------------------------===//
```

**Comment by lancep** [comment]
@Catfish-Man I think you want to do just a plain `benchmark` instead of `smoke benchmark`

**Comment by Catfish-Man** [comment]
@swift-ci please benchmark

**Comment by Catfish-Man** [comment]
Those numbers look MUCH more stable, excellent

**Comment by Catfish-Man** [comment]
@swift-ci please test and merge

**Comment by Catfish-Man** [comment]
@swift-ci please test and merge

**Comment by Catfish-Man** [comment]
@swift-ci please test and merge

**Pull request merged**